Problem: m0_servers sections is mandatory in CDF even for the client-only nodes

On the SAGE prototype cluster, there are server and client-only nodes. And if `m0_servers` section is not configured in the CDF for some client node, cfgen would fail with the error:

```
10:45 tkachuk1@sage-tier1-1:~$ hctl bootstrap --mkfs ~/sage-proto-cdf.yaml 
2020-11-17 10:46:01: Generating cluster configuration...cfgen: **ERROR** /home/users/jusers/tkachuk1/sage/sage-proto-cdf.yaml: Invalid cluster description

Error: Key m0_servers, expected by Dhall type:
List { io_disks : { data : List Text, meta_data : Optional Text }, runs_confd : Optional Bool }
is not present in YAML object:
data_iface: ib0
hostname: client-21
m0_clients:
  other: 80
  s3: 0
data_iface_type: o2ib

To learn more, run '/opt/seagate/cortx/hare/bin/../bin/cfgen --help-schema'.
```

To workaround the issue I specified the following for each client node:

```
    m0_servers: 
      - io_disks: 
          data: [] 
```

(Not sure whether it will work as needed though - whether it won't put io-services with empty disks list onto the client nodes. Will update on it.)

Looks like the workaround does not work after all:

```
12:55 seagate@sage-cmu:sage$ hctl bootstrap --mkfs ~tkachuk1/sage-proto-cdf.yaml
2020-11-17 12:55:35: Generating cluster configuration...Traceback (most recent call last):
  File "/opt/seagate/cortx/hare/bin/../bin/cfgen", line 1741, in <module>
    main()
  File "/opt/seagate/cortx/hare/bin/../bin/cfgen", line 133, in main
    validate_cluster_desc(cluster_desc)
  File "/opt/seagate/cortx/hare/bin/../bin/cfgen", line 231, in validate_cluster_desc
    validate_nodes_desc(desc['nodes'])
  File "/opt/seagate/cortx/hare/bin/../bin/cfgen", line 267, in validate_nodes_desc
    f"{name}: Either 'runs_confd' or 'io_disks.data' must be set"
AssertionError: client-21: Either 'runs_confd' or 'io_disks.data' must be set
```

--- a/cfgen/cfgen
+++ b/cfgen/cfgen
@@ -217,10 +217,14 @@ def enrich_cluster_desc(desc: Dict[str, Any], mock_p: bool) -> None:
         node['facts']['_memsize_MB'] = int(float(
             node['facts']['memorysize_mb']))
 
+        if 'm0_servers' not in node:
+            continue
+
         for m0d in node['m0_servers']:
             m0d.setdefault('runs_confd', False)
             m0d['_io_disks'] = get_disks(node['hostname'], mock_p,
                                          m0d['io_disks']['data'])
+
     if 'profiles' not in desc:
         sns_pools = [pool['name'] for pool in desc['pools']
                      if pool_type(pool) is PoolT.sns]
@@ -256,6 +260,12 @@ def validate_nodes_desc(nodes_desc: List[Dict[str, Any]]) -> None:
         assert node['facts'][ipaddr_key(iface)], \
             f"""{name}: {iface!r} interface has no IP address
 Make sure the value of data_iface in the CDF is correct."""
+
+        if 'm0_servers' not in node:
+            assert (node['m0_clients']['s3']
+                    + node['m0_clients']['other'] > 0), \
+                f'{name}: At least one client should be configured'
+            continue
 
         nr_confds = sum(1 for m0d in node['m0_servers'] if m0d['runs_confd'])
         assert nr_confds < 2, f'{name}: Too many confd services'
@@ -1688,7 +1698,10 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
     for node in cluster_desc['nodes']:
         node_id = ConfNode.build(conf, root_id, node)
         ctrl_id = None
-        if any(m0d['_io_disks'] for m0d in node['m0_servers']):
+        m0ds = []
+        if 'm0_servers' in node:
+            m0ds = node['m0_servers']
+        if any(m0d['_io_disks'] for m0d in m0ds):
             ctrl_id = ConfController.build(conf,
                                            ConfEnclosure.build(conf, rack_id),
                                            node_id)
@@ -1696,7 +1709,7 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
         ConfProcess.build(conf, node_id, node, ProcT.hax)
 
         confd_p = False
-        for m0d in node['m0_servers']:
+        for m0d in m0ds:
             ConfProcess.build(conf, node_id, node, ProcT.m0_server, m0d,
                               ctrl_id)
             if m0d['runs_confd']:

--- a/cfgen/cfgen
+++ b/cfgen/cfgen
@@ -1698,9 +1698,7 @@ def build_cluster(cluster_desc: Dict[str, Any]) -> Cluster:
     for node in cluster_desc['nodes']:
         node_id = ConfNode.build(conf, root_id, node)
         ctrl_id = None
-        m0ds = []
-        if 'm0_servers' in node:
-            m0ds = node['m0_servers']
+        m0ds = node.get('m0_servers', [])
         if any(m0d['_io_disks'] for m0d in m0ds):
             ctrl_id = ConfController.build(conf,
                                            ConfEnclosure.build(conf, rack_id),

--- a/cfgen/cfgen
+++ b/cfgen/cfgen
@@ -19,7 +19,7 @@ from typing import Any, Callable, Dict, Iterator, List, NamedTuple, Optional, \
 import yaml
 
 
-__version__ = '0.13'
+__version__ = '0.14'
 
 
 def parse_opts(argv):
@@ -75,7 +75,7 @@ nodes:
     data_iface: <str>  # name of network interface; e.g., eth1, eth1:c1
     data_iface_type: tcp|o2ib  # type of network interface;
                                # optional, defaults to "tcp"
-    m0_servers:
+    m0_servers:        # optional for client-only nodes
       - runs_confd: <bool>  # optional, defaults to false
         io_disks:
           meta_data: <str>  # device path for meta-data;

--- a/cfgen/dhall/types/ClusterDesc.dhall
+++ b/cfgen/dhall/types/ClusterDesc.dhall
@@ -28,7 +28,7 @@ let Node =
   { hostname : Text
   , data_iface : Text
   , data_iface_type: Optional ./Protocol.dhall
-  , m0_servers : List M0Server
+  , m0_servers : Optional (List M0Server)
   , m0_clients : { s3 : Natural, other : Natural }
   }
 

--- a/rfc/18/README.md
+++ b/rfc/18/README.md
@@ -30,7 +30,7 @@ nodes:
     data_iface: <str>  # name of network interface; e.g., eth1, eth1:c1
     data_iface_type: tcp|o2ib  # type of network interface;
                                # optional, defaults to "tcp"
-    m0_servers:
+    m0_servers:        # optional for client-only nodes
       - runs_confd: <bool>  # optional, defaults to false
         io_disks:
           meta_data: <str>  # device path for meta-data;

